### DISCUSSION TITLE
Next.js PKCE Woes

### DISCUSSION THREAD

**Discussion by robpetty** [supabase/supabase#20922] [Questions]
Hi.  I have been struggling, hopefully with just some configuration issues related to PKCE.  

The primary issue is the -with-supabase code generator code, works fine, IF you do not load an auth link outside the browser that created the initial email. Meaning for example, if I copy the URL from the email (say I have Outlook.com open in a different browser then where the Next.js application is) and paste it in the same browser where the application is, it works.  

If you click the link in Outlook however, you get errors like:

```
⨯ eL [AuthApiError]: invalid request: both auth code and code verifier should be non-empty
    at eX (/var/task/.next/server/chunks/967.js:1:110420)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async e1 (/var/task/.next/server/chunks/967.js:1:111362)
    at async eQ (/var/task/.next/server/chunks/967.js:1:111036)
    at async ao._exchangeCodeForSession (/var/task/.next/server/chunks/967.js:1:125100)
    at async /var/task/.next/server/chunks/967.js:1:130874 {
  __isAuthError: true,
  status: 400
}
``` 

After banging my head about this for quite some time I came across:

https://supabase.com/docs/guides/auth/server-side/email-based-auth-with-pkce-flow-for-ssr

So I created a new dummy app using:

`npx create-next-app@latest -e with-supabase nextjs-auth-test .`

And then created the:
`app/auth/confirm/route.ts`

As the doc stated.

The issue now is, I was unable to hit this endpoint as the emails that come from Supabase were not sending me to /auth/confirm but rather to /auth/callback.  (To note that is also the original problem, being able to click a confirm link in an email in a different browser instance).

I did update the confirmation email as below:

```
<hr />
Testing PKCE with Token for SSR
<br />
<p>Follow this link to confirm your user:</p>
<p>
  <a href="{{ localhost:3001 }}/auth/confirm?token_hash={{ .TokenHash }}&type=email">Confirm your email</a>
</p>
```

However, the URL would come like this in the email client:

https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcjozxfolkzckeguoumrb.supabase.co%2Fauth%2Fv1%2Fverify%3Ftoken%3Dpkce_562cd1376a24b0e16c5ba5afe47e48e80d20346cab3ef0dd76431957%26type%3Dsignup%26redirect_to%3Dhttp%3A%2F%2Flocalhost%3A3001%2Fauth%2Fcallback%2Fauth%2Fconfirm%3Ftoken_hash%3Dpkce_562cd1376a24b0e16c5ba5afe47e48e80d20346cab3ef0dd76431957%26type%3Demail&data=05%7C02%7C%7Ca8d258452c2945da482308dc22e57163%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638423616041473847%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=tUR6NW19Eh0y1v5RkRbBQuLEkLDCARvaSLRr%2FYptEz0%3D&reserved=0

Previously I tried to use what was suggested but {{ .SiteURL }} gave me my applications site (which is expended) hence the change I made.

Now, due to testing several times, below I realized it, I am no longer able to get the "Confirm your signup" emails that use the template as specified.

Now I am getting:

```
Confirm your email
Follow this link to confirm your email:

[Confirm your email address](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcjozxfolkzckeguoumrb.supabase.co%2Fauth%2Fv1%2Fverify%3Ftoken%3Dpkce_f407540295383b077a59e3246fd137a35481e64726fd32b271aab63d%26type%3Dsignup%26redirect_to%3Dhttp%3A%2F%2Flocalhost%3A3001%2Fauth%2Fcallback&data=05%7C02%7C%7Cce614ec2a2b245c00ff808dc22e6a898%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C638423621262795046%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=0nNtf3z3c8BP7dblOPmtVOsTlHpQTJnBxGDhIU0s6QE%3D&reserved=0)

Alternatively, enter the code: 673763
``` 

I am unable to find where this email is originating, its not one of the options in Supabase Email Templates.  I thought this might just be some safety check Supabase does if you keep registering the same email again and again, but even still, I copied the link in it, and tried to see if it would "free" the email up as before but nope.  Doesn't seem to do anything.  Also no idea where to even enter the code give.  I feel this has now caused it to act like a one time passcode instead of password based auth.

Does anyone have insight how to properly setup Signup in Supabase, that allows one to click a link in an email in another browser and it actually work?  Maybe I am incorrect about the page I linked earlier is allowing (or have done something wrong).

As to the error earlier, I have see people state simply do a try/catch and ignore, but surely that can't be the answer to this, can it?

Thanks!

**Comment by robpetty** [comment]
Just some more observations:

- The newer "Confirm your email" emails, that I can't tell where they are coming from, the link has token, not token_hash.  They also do not have type.
- The earlier emails DID have the properties I was expecting.  However, I think the issue at that time was self inflicted because the default Sign Up code has an email redirect to /auth/callback

I have since now tried using different email addresses (temp email service) on the actual proper application, and its confirm emails are now also sending with that code at the end and with a body that is NOT what I have configured in Supabase Email Templates.

**Comment by robpetty** [comment]
I filed a bug related to the template changing specifically here:

https://github.com/supabase/supabase/issues/20923

I found when I removed what I added and saved the template went back to working as before.  Granted my problems still remain but at least get back to what I had.

**Comment by robpetty** [comment]
Alright some more progress, did find in GoTrue where that other email is configured:

```
const defaultConfirmationMail = `<h2>Confirm your email</h2>

<p>Follow this link to confirm your email:</p>
<p><a href="{{ .ConfirmationURL }}">Confirm your email address</a></p>
<p>Alternatively, enter the code: {{ .Token }}</p>
```

At:
https://github.com/supabase/gotrue/blob/5ad703bddc6ec74f076cbe6ce1f942663343d47a/internal/mailer/template.go#L42

Looks to be the default confirmation email.  Not sure yet where a template parsing failure would cause usage of the default though yet. Maybe in the mailer.

https://github.com/supabase/gotrue/blob/5ad703bddc6ec74f076cbe6ce1f942663343d47a/internal/mailer/template.go#L109

Shows were the data is generated/parsed out and then passed to the Mailer in data.

**Comment by robpetty** [comment]
Alright, got it working.  As I worked through it, the template rendering breaking really through me off the track.

What is missing (and maybe is obvious to some) in the docs in their example is that .{{ .SiteURL }} is based on your URL Configuration -> Site URL.

So when your debugging, you can't use the example in the template:

```
<p>Follow this link to confirm your user:</p>
<a href="{{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash }}&type=email"
    >Confirm your email</a
  >
```

My earlier syntax was broke setting off the email going back to the default offered by GoTrue.  So fixing my syntax in the email:
```
  <a href="http://localhost:3002/auth/confirm?token_hash={{ .TokenHash }}&type=email"
    >Confirm your email2</a
  >
```

Fixed the actual issue, that with allowing email client links to confirm a user as one would expect (without having to be in same browser instance as the app your trying to sign up with).

Now in my email client, I can click link, it hits the /auth/confirm handler, and life is good.

Hopefully this helps anyone else suffering this.  Just remember you will have to update your template when your done testing.  I am not sure how to set additional variables in the template.

Would suggesting adding a debug endpoint be an option for site url?  Or a debug mode?  Not sure how to handle this in a seamless fashion.

**Reply by rokanost** [reply]
Thank you, this was so damn confusing. I even wanted to drop using supabase auth.
